kmod/src/data.c

```diff
@@ -1274,21 +1281,47 @@ int scoutfs_data_move_blocks(struct inode *from, u64 from_off,
 
 			to_start = to_iblock + (from_start - from_iblock);
 
-			/* insert the new, fails if it overlaps */
-			ret = scoutfs_ext_insert(sb, &data_ext_ops, &to_args,
-						 to_start, len,
-						 map, ext.flags);
+			if (is_stage) {
+				struct scoutfs_extent to_ext;
+
+				ret = scoutfs_ext_next(sb, &data_ext_ops, &to_args,
+						       to_iblock, 1, &to_ext);
+				if (ret)
+					break;
+
+				if (!scoutfs_ext_inside(to_start, len, &to_ext) ||
+				    !(to_ext.flags & SEF_OFFLINE)) {
+					ret = -EINVAL;
+					break;
+				}
+
+				ret = scoutfs_ext_set(sb, &data_ext_ops, &to_args,
+							 to_start, len,
+							 map, ext.flags);
+			} else {
+				/* insert the new, fails if it overlaps */
+				ret = scoutfs_ext_insert(sb, &data_ext_ops, &to_args,
+							 to_start, len,
+							 map, ext.flags);
+			}
 			if (ret < 0)
 				break;
 
 			/* remove the old, possibly splitting */
 			ret = scoutfs_ext_set(sb, &data_ext_ops, &from_args,
 					      from_start, len, 0, 0);
 			if (ret < 0) {
-				/* remove inserted new on err */
-				err = scoutfs_ext_remove(sb, &data_ext_ops,
-							 &to_args, to_start,
-							 len);
+				if (is_stage) {
+					/* re-mark dest range as offline */
+					err = scoutfs_ext_set(sb, &data_ext_ops, &to_args,
+							      to_start, len,
+							      0, ext.flags | SEF_OFFLINE);
```

Comment:
I don't think we should assume that ext.flags | offline == to_ext.flags.  That might not even be the case today if we try to move unwritten extents, and certainly might not be the case if we add weird flags later.  Can we move the to_ext from being local to its block so that we can use to_ext.flags here to restore the original flags?    (maybe call it off_ext?)

Reply:
ah yes that's the right thing to do.